components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -186,12 +186,28 @@ fn update_langid(
     entry: &LanguageIdentifier,
     langid: &mut LanguageIdentifier,
 ) -> CanonicalizationResult {
-    if langid.language.is_empty() {
+    let mut modified: bool = false;
+
+    if langid.language.is_empty() && !entry.language.is_empty() {
         langid.language = entry.language;
+        modified = true;
+    }
+
+    if langid.script.is_none() && entry.script.is_some() {
+        langid.script = entry.script;
+        modified = true;
     }
-    langid.script = langid.script.or(entry.script);
-    langid.region = langid.region.or(entry.region);
-    CanonicalizationResult::Modified
+
+    if langid.region.is_none() && entry.region.is_some() {
+        langid.region = entry.region;
+        modified = true;
+    }
+
+    if modified {
+        return CanonicalizationResult::Modified;
+    }
+
+    CanonicalizationResult::Unmodified
```

Comment:
`nit (n-b)`: I'm surprised `cargo clippy` didn't warn you, Rust prefers matching branches and implicit `return`:

```rust
if modified {
    CanonicalizationResult::Modified
} else {
    CanonicalizationResult::Unmodified
}
```